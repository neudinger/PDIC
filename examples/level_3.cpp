/*
 * File: level_3.cpp
 * Project: examples
 * File Created: Sunday, 25th July 2021 8:04:25 pm
 * Author: kbarre (kevin.barre@epitech.eu)
 * -----
 * Last Modified: Sunday, 25th July 2021 8:04:55 pm
 * Modified By: kbarre (kevin.barre@epitech.eu>)
 * -----
 * Licenses: EUPL
 * -----
 * Copyright 2021 - 2021 neudinger
 */

#pragma pdi on

typedef struct Structure1
{
    int id1;          // 4 Bytes
    int id2;          // 4 Bytes
    char name;        // 1 Bytes
    char c;           // 1 Bytes
    float percentage; // 4 Bytes
    template <typename T>
    operator T *()
    {
        return new T();
    }
} St1;

struct structure2
{
    int id1;          // 4 Bytes
    char name;        // 1 Bytes
    int id2;          // 4 Bytes
    char c;           // 1 Bytes
    float percentage; // 4 Bytes
    St1 *val = new St1();
} __attribute__((packed));


typedef struct // <- warning: will be printed cause of Anonymous struct used but alias will be used
{
    struct structure2 st2;
    St1 my_st1;
} St3;

struct structure2 dd;
St3 global_st3;

#pragma pdi off