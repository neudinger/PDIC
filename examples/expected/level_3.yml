structs:
 Structure1:
  type: record
  name: Structure1
  alias: [St1]
  fieldsize: 5
  buffersize: 16
  packed: false
  members:
   c:
    offset: 9   
    type: char
   id1:   
    type: int
   id2:
    offset: 4   
    type: int
   name:
    offset: 8   
    type: char
   percentage:
    offset: 12   
    type: float
 structure2:
  type: record
  name: structure2
  fieldsize: 6
  buffersize: 22
  packed: true
  members:
   c:
    offset: 9   
    type: char
   id1:   
    type: int
   id2:
    offset: 5   
    type: int
   name:
    offset: 4   
    type: char
   percentage:
    offset: 10   
    type: float
   val: { type: pointer, subtype: St1, offset: 14 }
 St3:
  type: record
  name: St3
  alias: [St3]
  fieldsize: 2
  buffersize: 40
  packed: false
  members:
   my_st1:
    offset: 24
    type: record
    name: Structure1
    alias: [St1]
    fieldsize: 5
    buffersize: 16
    packed: false
    members:
     c:
      offset: 9     
      type: char
     id1:     
      type: int
     id2:
      offset: 4     
      type: int
     name:
      offset: 8     
      type: char
     percentage:
      offset: 12     
      type: float

   st2:
    type: record
    name: structure2
    fieldsize: 6
    buffersize: 22
    packed: true
    members:
     c:
      offset: 9     
      type: char
     id1:     
      type: int
     id2:
      offset: 5     
      type: int
     name:
      offset: 4     
      type: char
     percentage:
      offset: 10     
      type: float
     val: { type: pointer, subtype: St1, offset: 14 }

data:
 dd:
  type: record
  name: structure2
  fieldsize: 6
  buffersize: 22
  packed: true
  members:
   c:
    offset: 9   
    type: char
   id1:   
    type: int
   id2:
    offset: 5   
    type: int
   name:
    offset: 4   
    type: char
   percentage:
    offset: 10   
    type: float
   val: { type: pointer, subtype: St1, offset: 14 }

 global_st3:
  type: record
  name: St3
  alias: [St3]
  fieldsize: 2
  buffersize: 40
  packed: false
  members:
   my_st1:
    offset: 24
    type: record
    name: Structure1
    alias: [St1]
    fieldsize: 5
    buffersize: 16
    packed: false
    members:
     c:
      offset: 9     
      type: char
     id1:     
      type: int
     id2:
      offset: 4     
      type: int
     name:
      offset: 8     
      type: char
     percentage:
      offset: 12     
      type: float

   st2:
    type: record
    name: structure2
    fieldsize: 6
    buffersize: 22
    packed: true
    members:
     c:
      offset: 9     
      type: char
     id1:     
      type: int
     id2:
      offset: 5     
      type: int
     name:
      offset: 4     
      type: char
     percentage:
      offset: 10     
      type: float
     val: { type: pointer, subtype: St1, offset: 14 }


